In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import json
from pandas.io.json import json_normalize

In [2]:
#We are working with processed world bank project data
with open('data/world_bank_projects.json') as data_file:    
    world_bank_data = json.load(data_file) #Load it up for analysis

In [3]:
#Load and check out the columns
json_df = pd.DataFrame(world_bank_data)
json_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

In [4]:
#There are a lot of data, so let's remove some of them
json_df_less = json_df[['_id', 'boardapprovaldate',
                        'countryname', 'countryshortname',
                        'mjtheme_namecode', 'project_name',
                        'totalamt']]
json_df_less

_id     boardapprovaldate  \
0    {'$oid': '52b213b38594d8a2be17c780'}  2013-11-12T00:00:00Z   
1    {'$oid': '52b213b38594d8a2be17c781'}  2013-11-04T00:00:00Z   
2    {'$oid': '52b213b38594d8a2be17c782'}  2013-11-01T00:00:00Z   
3    {'$oid': '52b213b38594d8a2be17c783'}  2013-10-31T00:00:00Z   
4    {'$oid': '52b213b38594d8a2be17c784'}  2013-10-31T00:00:00Z   
5    {'$oid': '52b213b38594d8a2be17c785'}  2013-10-31T00:00:00Z   
6    {'$oid': '52b213b38594d8a2be17c786'}  2013-10-29T00:00:00Z   
7    {'$oid': '52b213b38594d8a2be17c787'}  2013-10-29T00:00:00Z   
8    {'$oid': '52b213b38594d8a2be17c788'}  2013-10-29T00:00:00Z   
9    {'$oid': '52b213b38594d8a2be17c789'}  2013-10-29T00:00:00Z   
10   {'$oid': '52b213b38594d8a2be17c78a'}  2013-10-25T00:00:00Z   
11   {'$oid': '52b213b38594d8a2be17c78b'}  2013-10-25T00:00:00Z   
12   {'$oid': '52b213b38594d8a2be17c78c'}  2013-10-24T00:00:00Z   
13   {'$oid': '52b213b38594d8a2be17c78d'}  2013-10-22T00:00:00Z   
14   {'$oid': '52b213b38594d8a2be17c78e'}  2013-10-22T00:00:00Z   
15   {'$oid': '52b213b38594d8a2be17c78f'}  2013-10-17T00:00:00Z   
16   {'$oid': '52b213b38594d8a2be17c790'}  2013-10-17T00:00:00Z   
17   {'$oid': '52b213b38594d8a2be17c791'}  2013-10-16T00:00:00Z   
18   {'$oid': '52b213b38594d8a2be17c792'}  2013-10-16T00:00:00Z   
19   {'$oid': '52b213b38594d8a2be17c793'}  2013-10-10T00:00:00Z   
20   {'$oid': '52b213b38594d8a2be17c794'}  2013-10-09T00:00:00Z   
21   {'$oid': '52b213b38594d8a2be17c795'}  2013-10-07T00:00:00Z   
22   {'$oid': '52b213b38594d8a2be17c796'}  2013-10-07T00:00:00Z   
23   {'$oid': '52b213b38594d8a2be17c797'}  2013-10-03T00:00:00Z   
24   {'$oid': '52b213b38594d8a2be17c798'}  2013-09-30T00:00:00Z   
25   {'$oid': '52b213b38594d8a2be17c799'}  2013-09-30T00:00:00Z   
26   {'$oid': '52b213b38594d8a2be17c79a'}  2013-09-30T00:00:00Z   
27   {'$oid': '52b213b38594d8a2be17c79b'}  2013-09-30T00:00:00Z   
28   {'$oid': '52b213b38594d8a2be17c79c'}  2013-09-30T00:00:00Z   
29   {'$oid': '52b213b38594d8a2be17c79d'}  2013-09-27T00:00:00Z   
..                                    ...                   ...   
470  {'$oid': '52b213b38594d8a2be17c956'}  2012-09-20T00:00:00Z   
471  {'$oid': '52b213b38594d8a2be17c957'}  2012-09-20T00:00:00Z   
472  {'$oid': '52b213b38594d8a2be17c958'}  2012-09-13T00:00:00Z   
473  {'$oid': '52b213b38594d8a2be17c959'}  2012-09-13T00:00:00Z   
474  {'$oid': '52b213b38594d8a2be17c95a'}  2012-09-11T00:00:00Z   
475  {'$oid': '52b213b38594d8a2be17c95b'}  2012-09-11T00:00:00Z   
476  {'$oid': '52b213b38594d8a2be17c95c'}  2012-09-11T00:00:00Z   
477  {'$oid': '52b213b38594d8a2be17c95d'}  2012-09-11T00:00:00Z   
478  {'$oid': '52b213b38594d8a2be17c95e'}  2012-09-11T00:00:00Z   
479  {'$oid': '52b213b38594d8a2be17c95f'}  2012-09-10T00:00:00Z   
480  {'$oid': '52b213b38594d8a2be17c960'}  2012-09-10T00:00:00Z   
481  {'$oid': '52b213b38594d8a2be17c961'}  2012-09-06T00:00:00Z   
482  {'$oid': '52b213b38594d8a2be17c962'}  2012-09-06T00:00:00Z   
483  {'$oid': '52b213b38594d8a2be17c963'}  2012-09-06T00:00:00Z   
484  {'$oid': '52b213b38594d8a2be17c964'}  2012-09-06T00:00:00Z   
485  {'$oid': '52b213b38594d8a2be17c965'}  2012-09-05T00:00:00Z   
486  {'$oid': '52b213b38594d8a2be17c966'}  2012-09-01T00:00:00Z   
487  {'$oid': '52b213b38594d8a2be17c967'}  2012-08-30T00:00:00Z   
488  {'$oid': '52b213b38594d8a2be17c968'}  2012-08-30T00:00:00Z   
489  {'$oid': '52b213b38594d8a2be17c969'}  2012-08-30T00:00:00Z   
490  {'$oid': '52b213b38594d8a2be17c96a'}  2012-08-29T00:00:00Z   
491  {'$oid': '52b213b38594d8a2be17c96b'}  2012-08-24T00:00:00Z   
492  {'$oid': '52b213b38594d8a2be17c96c'}  2012-08-21T00:00:00Z   
493  {'$oid': '52b213b38594d8a2be17c96d'}  2012-08-21T00:00:00Z   
494  {'$oid': '52b213b38594d8a2be17c96e'}  2012-08-17T00:00:00Z   
495  {'$oid': '52b213b38594d8a2be17c96f'}  2012-08-10T00:00:00Z   
496  {'$oid': '52b213b38594d8a2be17c970'}  2012-08-09T00:00:00Z   
497  {'$oid': '52b213b38594d8a2be17c971'}  2012-08-03T00:00:00Z   
498  {'$oid': '52b

#### Let us find the 10 countries with the most projects

In [5]:
#Top ten countries with most projects
json_df_less['countryname'].value_counts().head(10)

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
People's Republic of Bangladesh    12
Kingdom of Morocco                 12
Nepal                              12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

#### 10 Major Project themes

In [6]:
#Top ten major project themes (Some themes are counted multiple times in a single project)
wb_df_norm = json_normalize(world_bank_data, 'mjtheme_namecode')
wb_df_norm['name'].value_counts().head(10)


Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

###### From the output above, we can see that there are 122 missing values.

In [7]:
wb_df_norm['code'].value_counts().head(10)

11    250
10    216
8     210
2     199
6     168
4     146
7     130
5      77
9      50
1      38
Name: code, dtype: int64

###### We can see that the output above shows that just the codes are available, only the names are missing

#### Create a dataframe with the missing names filled in.

In [8]:
#Manually created dictionary for comparison purposes
theme_code_dict = {'1':'Economic management',
             '2':'Public sector governance',
             '3':'Rule of law',
             '4':'Financial and private sector development',
             '5':'Trade and integration',
             '6':'Social protection and risk management',
             '7':'Social dev/gender/inclusion',
             '8':'Human development',
             '9':'Urban development',
             '10':'Rural development',
             '11':'Environment and natural resources management',
            }
#Print dict
theme_code_dict

{'1': 'Economic management',
 '2': 'Public sector governance',
 '3': 'Rule of law',
 '4': 'Financial and private sector development',
 '5': 'Trade and integration',
 '6': 'Social protection and risk management',
 '7': 'Social dev/gender/inclusion',
 '8': 'Human development',
 '9': 'Urban development',
 '10': 'Rural development',
 '11': 'Environment and natural resources management'}

In [9]:
#Non-manual version
json_norm = json_normalize(world_bank_data, 'mjtheme_namecode') #Split column into own table
jn_no_empty = json_norm[json_norm.name != ''].set_index('code')['name'] #Remove empty names 
theme_code_dict = jn_no_empty.drop_duplicates().to_dict() #Drop duplicates and convert to dictionary
theme_code_dict #Print dict

{'8': 'Human development',
 '1': 'Economic management',
 '6': 'Social protection and risk management',
 '5': 'Trade and integration',
 '2': 'Public sector governance',
 '11': 'Environment and natural resources management',
 '7': 'Social dev/gender/inclusion',
 '4': 'Financial and private sector development',
 '10': 'Rural development',
 '9': 'Urban development',
 '3': 'Rule of law'}

In [10]:
#Function to replace mjtheme name using code from dictionary
def replace_empty(theme_dict_list):
    for theme_dict in theme_dict_list:
        if theme_dict['name'] == '':
            theme_dict['name'] = theme_code_dict[theme_dict['code']]
    return theme_dict

In [11]:
filled_json_df = json_df.copy(deep=True)
filled_json_df['mjtheme_namecode'].apply(replace_empty).head(10)

0    {'code': '11', 'name': 'Environment and natura...
1    {'code': '6', 'name': 'Social protection and r...
2    {'code': '6', 'name': 'Social protection and r...
3    {'code': '7', 'name': 'Social dev/gender/inclu...
4    {'code': '4', 'name': 'Financial and private s...
5    {'code': '6', 'name': 'Social protection and r...
6    {'code': '4', 'name': 'Financial and private s...
7           {'code': '8', 'name': 'Human development'}
8    {'code': '7', 'name': 'Social dev/gender/inclu...
9    {'code': '2', 'name': 'Public sector governance'}
Name: mjtheme_namecode, dtype: object

In [12]:
#Top ten major project themes (Some themes are counted multiple times in a single project)
wb_df_clean = json_normalize(world_bank_data, 'mjtheme_namecode')
wb_df_clean['name'].value_counts().head(10)

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Name: name, dtype: int64

In [16]:
json_df_less

_id     boardapprovaldate  \
0    {'$oid': '52b213b38594d8a2be17c780'}  2013-11-12T00:00:00Z   
1    {'$oid': '52b213b38594d8a2be17c781'}  2013-11-04T00:00:00Z   
2    {'$oid': '52b213b38594d8a2be17c782'}  2013-11-01T00:00:00Z   
3    {'$oid': '52b213b38594d8a2be17c783'}  2013-10-31T00:00:00Z   
4    {'$oid': '52b213b38594d8a2be17c784'}  2013-10-31T00:00:00Z   
5    {'$oid': '52b213b38594d8a2be17c785'}  2013-10-31T00:00:00Z   
6    {'$oid': '52b213b38594d8a2be17c786'}  2013-10-29T00:00:00Z   
7    {'$oid': '52b213b38594d8a2be17c787'}  2013-10-29T00:00:00Z   
8    {'$oid': '52b213b38594d8a2be17c788'}  2013-10-29T00:00:00Z   
9    {'$oid': '52b213b38594d8a2be17c789'}  2013-10-29T00:00:00Z   
10   {'$oid': '52b213b38594d8a2be17c78a'}  2013-10-25T00:00:00Z   
11   {'$oid': '52b213b38594d8a2be17c78b'}  2013-10-25T00:00:00Z   
12   {'$oid': '52b213b38594d8a2be17c78c'}  2013-10-24T00:00:00Z   
13   {'$oid': '52b213b38594d8a2be17c78d'}  2013-10-22T00:00:00Z   
14   {'$oid': '52b213b38594d8a2be17c78e'}  2013-10-22T00:00:00Z   
15   {'$oid': '52b213b38594d8a2be17c78f'}  2013-10-17T00:00:00Z   
16   {'$oid': '52b213b38594d8a2be17c790'}  2013-10-17T00:00:00Z   
17   {'$oid': '52b213b38594d8a2be17c791'}  2013-10-16T00:00:00Z   
18   {'$oid': '52b213b38594d8a2be17c792'}  2013-10-16T00:00:00Z   
19   {'$oid': '52b213b38594d8a2be17c793'}  2013-10-10T00:00:00Z   
20   {'$oid': '52b213b38594d8a2be17c794'}  2013-10-09T00:00:00Z   
21   {'$oid': '52b213b38594d8a2be17c795'}  2013-10-07T00:00:00Z   
22   {'$oid': '52b213b38594d8a2be17c796'}  2013-10-07T00:00:00Z   
23   {'$oid': '52b213b38594d8a2be17c797'}  2013-10-03T00:00:00Z   
24   {'$oid': '52b213b38594d8a2be17c798'}  2013-09-30T00:00:00Z   
25   {'$oid': '52b213b38594d8a2be17c799'}  2013-09-30T00:00:00Z   
26   {'$oid': '52b213b38594d8a2be17c79a'}  2013-09-30T00:00:00Z   
27   {'$oid': '52b213b38594d8a2be17c79b'}  2013-09-30T00:00:00Z   
28   {'$oid': '52b213b38594d8a2be17c79c'}  2013-09-30T00:00:00Z   
29   {'$oid': '52b213b38594d8a2be17c79d'}  2013-09-27T00:00:00Z   
..                                    ...                   ...   
470  {'$oid': '52b213b38594d8a2be17c956'}  2012-09-20T00:00:00Z   
471  {'$oid': '52b213b38594d8a2be17c957'}  2012-09-20T00:00:00Z   
472  {'$oid': '52b213b38594d8a2be17c958'}  2012-09-13T00:00:00Z   
473  {'$oid': '52b213b38594d8a2be17c959'}  2012-09-13T00:00:00Z   
474  {'$oid': '52b213b38594d8a2be17c95a'}  2012-09-11T00:00:00Z   
475  {'$oid': '52b213b38594d8a2be17c95b'}  2012-09-11T00:00:00Z   
476  {'$oid': '52b213b38594d8a2be17c95c'}  2012-09-11T00:00:00Z   
477  {'$oid': '52b213b38594d8a2be17c95d'}  2012-09-11T00:00:00Z   
478  {'$oid': '52b213b38594d8a2be17c95e'}  2012-09-11T00:00:00Z   
479  {'$oid': '52b213b38594d8a2be17c95f'}  2012-09-10T00:00:00Z   
480  {'$oid': '52b213b38594d8a2be17c960'}  2012-09-10T00:00:00Z   
481  {'$oid': '52b213b38594d8a2be17c961'}  2012-09-06T00:00:00Z   
482  {'$oid': '52b213b38594d8a2be17c962'}  2012-09-06T00:00:00Z   
483  {'$oid': '52b213b38594d8a2be17c963'}  2012-09-06T00:00:00Z   
484  {'$oid': '52b213b38594d8a2be17c964'}  2012-09-06T00:00:00Z   
485  {'$oid': '52b213b38594d8a2be17c965'}  2012-09-05T00:00:00Z   
486  {'$oid': '52b213b38594d8a2be17c966'}  2012-09-01T00:00:00Z   
487  {'$oid': '52b213b38594d8a2be17c967'}  2012-08-30T00:00:00Z   
488  {'$oid': '52b213b38594d8a2be17c968'}  2012-08-30T00:00:00Z   
489  {'$oid': '52b213b38594d8a2be17c969'}  2012-08-30T00:00:00Z   
490  {'$oid': '52b213b38594d8a2be17c96a'}  2012-08-29T00:00:00Z   
491  {'$oid': '52b213b38594d8a2be17c96b'}  2012-08-24T00:00:00Z   
492  {'$oid': '52b213b38594d8a2be17c96c'}  2012-08-21T00:00:00Z   
493  {'$oid': '52b213b38594d8a2be17c96d'}  2012-08-21T00:00:00Z   
494  {'$oid': '52b213b38594d8a2be17c96e'}  2012-08-17T00:00:00Z   
495  {'$oid': '52b213b38594d8a2be17c96f'}  2012-08-10T00:00:00Z   
496  {'$oid': '52b213b38594d8a2be17c970'}  2012-08-09T00:00:00Z   
497  {'$oid': '52b213b38594d8a2be17c971'}  2012-08-03T00:00:00Z   
498  {'$oid': '52b